RD-7117 Stage system tests failed on deployments_spec

leon-good-life opened the pull request:
<!--- Provide JIRA issue ID and a general summary of your changes in the Title above -->
<!--- e.g. "RD-1234 Improve styling in SuperComponent" -->
<!--- Note that the title is used as a commit message after merging the PR -->

## Description
Jira ticket: RD-7117
<!--- Describe your changes to help reviewers understand the details. -->

## Screenshots / Videos

Before:
<img width="1279" alt="Screenshot 2023-03-23 at 12 24 05" src="https://user-images.githubusercontent.com/17104212/227174666-0302bef9-49a8-4373-b56d-71a4ad90a952.png">

After:
<img width="1390" alt="Screenshot 2023-03-23 at 12 25 26" src="https://user-images.githubusercontent.com/17104212/227175027-da986c34-53fc-4149-a4df-67a3cd3cc953.png">

<!--- If applicable, add screenshot(s) or video(s) describing the changes you made. -->
<!--- Consider adding comparison screenshot(s) - before and after the change. -->
<!--- If not applicable, just write “N/A” in this section. -->

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I followed the [UI Testing Guidelines](https://cloudifysource.atlassian.net/l/cp/udCVfvrx) and verified that all tests/checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests
`deployments_spec` - ✅ passed:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/4794/pipeline/

`deployments_spec` - ✅ passed:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/4793/pipeline
<!--- If applicable, describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If applicable, provide a link to system tests, e.g: -->
<!--- [#1039](https://jenkins.cloudify.co/.../1039) (2021-08-06 07:05:50) --> 
<!--- If not applicable, just write “N/A” in this section. -->

## Documentation
N/A
<!--- If applicable, describe how you documented or plan to document your changes. -->
<!--- Check UI documentation guidelines - https://cloudifysource.atlassian.net/l/c/EYWJ0XfB - for details. -->
<!--- Provide links to JIRA issues, other PRs or related documentation pages. -->
<!--- If not applicable, just write “N/A” in this section. -->

Vorbert-Kruk commented on app/widgets/common/deployments/DeploymentActionsMenu.utils.ts:
```diff
@@ -4,7 +4,7 @@ import type { MenuItem } from './DeploymentActionsMenu.consts';
 import { permissions, actions } from './DeploymentActionsMenu.consts';
 
 function hasEnvironmentLabel(deploymentLabels: Label[]) {
-    return deploymentLabels.some(deploymentLabel => {
+    return deploymentLabels?.some(deploymentLabel => {
         return deploymentLabel.key === 'csys-obj-type' && deploymentLabel.value === 'environment';
     });
 }
```
WDYT about the following modification?

```suggestion
function hasEnvironmentLabel(deploymentLabels: Label[] = []) {
    return deploymentLabels.some(deploymentLabel => {
        return deploymentLabel.key === 'csys-obj-type' && deploymentLabel.value === 'environment';
    });
}
```

leon-good-life replied:
[b6662f3](https://github.com/cloudify-cosmo/cloudify-stage/pull/2496/commits/b6662f38f932f90939227d8e6e09984dbe6cf7d5)

qooban commented on app/widgets/common/deployments/DeploymentActionsMenu.utils.ts:
```diff
@@ -4,7 +4,7 @@ import type { MenuItem } from './DeploymentActionsMenu.consts';
 
 function hasEnvironmentLabel(deploymentLabels: Label[]) {
-    return deploymentLabels.some(deploymentLabel => {
+    return deploymentLabels?.some(deploymentLabel => {
```
What was the actual error? What was the value of `deploymentLabels` passed to `hasEnvironmentLabel` function?

According to the type - `deploymentLabels` are supposed to be an array of `Label` objects, so adding `?` seems to me like a way to confuse reader who would expect `deploymentLabels` to not be undefined/null.

leon-good-life replied:
Here is the error:
<img width="1039" alt="Screenshot 2023-03-23 at 13 09 10" src="https://user-images.githubusercontent.com/17104212/227185726-edefa89a-ace5-40d8-8e4d-33842ed36bcc.png">

The value of `deploymentLabels` was `undefined` that passed to `hasEnvironmentLabel`:
<img width="1060" alt="Screenshot 2023-03-23 at 13 11 50" src="https://user-images.githubusercontent.com/17104212/227186356-08a8e973-ae42-41a9-9302-9796071ea209.png">

qooban replied:
Hmm... Interesting. Could you check what is the actual response from manager API for that problematic deployment? (on /deployments endpoint) Do we receive nullish value on `labels` field?

qooban replied:
@leon-good-life @Vorbert-Kruk 

I checked that by myself. Will share my findings here:

1. In this specific place in test we were using mocked data for deployments endpoint response. This mocked data lacked `labels` property. That was direct reason for the problem and it should be fixed in the first place.
2. When deployment does not have labels defined, it responds with `labels: []`, so there's no need to do any workarounds.

Having that said and knowing that Leon is on vacation, I'm taking over RD-7117 and I'll provide fixes for `master` and `7.0.0-build` branches.

Vorbert-Kruk commented:
I've approved due to a change that is fixing the issue
But @qooban as `deploymentLabels` could not exist in the component, it would be beneficial if we'd adjust the typing used within component accordingly
I don't know if the current PR is a good place for that or if a new ticket should be created (maybe while working on ([RND-21](https://cloudifysource.atlassian.net/browse/RND-21)) I could tackle it)
But as the issue seems to have a deeper root cause, I've started to think about possible approaches to it
WDYT? 🤔

leon-good-life commented:
I run the test (`deployments_spec`) again after addressing review comment and it passed - ✅ passed:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/4794/pipeline/

leon-good-life commented:
I'm encountering 502 bad gateway when I'm installing version 7.0.0. So I unable to test it on 7.0.0. I guess the test is failing also on 7.0.0. So I will make a PR to `7.0.0-build`.
